### ISSUE TITLE
JSON format: file names may be stored in relative form.

### ISSUE THREAD

**Issue by xvolks** [ryouchinsa/Rectlabel-support#8]
I use RectLabel to label images on my Mac. The resulting files are then copied to big server with plenty of CPUs and GPUs running under Linux.

The issue is that I need to modify each JSON file to match the file system of the Linux box.
If the path was relative between JSON and image file, my job would be easier 😉

A simple checkbox in preferences may do the trick.

**Comment by ryouchinsa** [comment]
Thanks for using RectLabel.
Are you modifying "image_path" in the JSON format?

"image_path" is just for reminding you of for which image the label file was created.
So that RectLabel does not use this path when reading the image and the label file.

When you open a folder, RectLabel could know the image folder path.
The "annotations" folder has to be in the path "image_folder_path/annotations".
```xml
{
  "objects" : [
    {
      "label" : "face-x_45-y45",
      "x_y_w_h" : [
        238,
        116,
        135,
        135
      ]
    }
  ],
  "image_path" : "\/Users\/ryo\/Desktop\/image\/test1.jpg",
  "image_w_h" : [
    640,
    960
  ]
}
```

**Comment by xvolks** [comment]
I wrote my own script to load JSON files for the training data based on [this script](https://github.com/yhenon/keras-frcnn/blob/master/keras_frcnn/simple_parser.py).

The `image_path` value is used by the script to read the actual image for the training set.
As the label phase takes place on my Mac and the training takes places on a Linux box, the paths are very different. Since I wrote my own script, I can take care of those differences and remove the Mac path to replace it with the Linux path on the fly.

**Comment by ryouchinsa** [comment]
Thanks for telling us the use case.
Based on the PASCAL VOC format, the JSON format also should have "folder" and "filename".

```xml
{
  "filename" : "test1.jpg",
  "folder" : "image",
  "image_w_h" : [
    640,
    960
  ],
  "objects" : [
    {
      "label" : "face-x_45-y45",
      "x_y_w_h" : [
        263,
        187,
        60,
        54
      ]
    }
  ]
}
```

This would fit with how "keras-frcnn" reads both images and annotations in their application code.
If you could agree with this idea, we would update RectLabel again.

**Comment by xvolks** [comment]
I have added this parser to keras-frcnn : json_parser.py in keras_frcnn package with a small modification in train_frcnn.py.
It works for my use case.

2 lines more in train_frcnn.py 
elif options.parser == 'simple':
    from keras_frcnn.simple_parser import get_data
+ elif options.parser == 'json':
+     from keras_frcnn.json_parser import get_data
else:
    raise ValueError("Command line option parser must be one of 'pascal_voc' or 'simple'")

File json_parser.py
import json as js
import os
import cv2
import numpy as np

def get_data(input_path):
    found_bg = False
    all_imgs = {}

    classes_count = {}

    class_mapping = {}

    all_data = []

    visualise = True

    for root, dirs, files in os.walk(input_path):
        for name in files:
            if not name.endswith(".json"):
                continue
            js_file = os.path.join(root, name)
            print(js_file)
            with open(js_file, 'r') as f:
                content = js.load(f)
                filename = content['image_path']
                filename = os.path.join(input_path, os.path.basename(filename))
                if filename not in all_imgs:
                    all_imgs[filename] = {}
                    rows = content['image_w_h'][1]
                    cols = content['image_w_h'][0]
                    #img = cv2.imread(filename)
                    #(rows, cols) = img.shape[:2]
                    all_imgs[filename]['filepath'] = filename
                    all_imgs[filename]['width'] = cols
                    all_imgs[filename]['height'] = rows
                    all_imgs[filename]['bboxes'] = []
                    if np.random.randint(0, 6) > 0:
                        all_imgs[filename]['imageset'] = 'trainval'
                    else:
                        all_imgs[filename]['imageset'] = 'test'

                objects = content['objects']
                for object in objects:
                    class_name = object['label']
                    coords = object['x_y_w_h']
                    if class_name not in classes_count:
                        classes_count[class_name] = 1
                    else:
                        classes_count[class_name] += 1

                    if class_name not in class_mapping:
                        if class_name == 'bg' and found_bg == False:
                            print(
                            'Found class name with special name bg. Will be treated as a background region (this is usually for hard negative mining).')
                            found_bg = True
                        class_mapping[class_name] = len(class_mapping)

                    all_imgs[filename]['bboxes'].append(
                        {'class': class_name, 'x1': coords[0], 'y1': coords[1], 'x2': coords[0]+coords[2], 'y2': coords[1] + coords[3]})

                pass

        for key in all_imgs:
            all_data.append(all_imgs[key])

    # make sure the bg class is last in the list
    if found_bg:
        if class_mapping['bg'] != len(class_mapping) - 1:
            key_to_switch = [key for key in class_mapping.keys() if class_mapping[key] == len(class_mapping) - 1][0]
            val_to_switch = class_mapping['bg']
            class_mapping['bg'] = len(class_mapping) - 1
            class_mapping[key_to_switch] = val_to_switch

    return all_data, classes_count, class_mapping

Best regards.

**Comment by ryouchinsa** [comment]
In your code, with the new update in a few days, you have to change like below.

`filename = content['image_path']`
`filename = os.path.join(input_path, os.path.basename(filename))`
=>
`filename = content['filename']`
`filename = os.path.join(input_path, filename)`

**Comment by xvolks** [comment]
Thank you.

[Kato Sensei wrote.](http://eej.free.fr/kato%20calligraphie.jpg)

**Comment by ryouchinsa** [comment]
The new update was released.
RectLabel 1.32
Based on the PASCAL VOC format, the JSON format also should have "folder" and "filename".
